File: app/src/main/java/com/ptithcm/thuan6420/basecleanarchitecture/ui/login/UserController.java
package com.ptithcm.thuan6420.basecleanarchitecture.ui.login;

import android.view.View;

import com.ptithcm.thuan6420.basecleanarchitecture.R;
import com.ptithcm.thuan6420.basecleanarchitecture.data.repositories.UserRepository;
import com.ptithcm.thuan6420.basecleanarchitecture.ui.components.UiController;
import com.ptithcm.thuan6420.basecleanarchitecture.ui.dialogs.DialogController;
import com.ptithcm.thuan6420.basecleanarchitecture.ui.util.ConstantMessage;


public class UserController {
    private User model;
    private LoginFragment loginFragment;
    private RegisterFragment registerFragment;

    public UserController(LoginFragment loginFragment) {
        this.model = new User();
        this.loginFragment = loginFragment;
    }

    public UserController(RegisterFragment registerFragment) {
        this.model = new User();
        this.registerFragment = registerFragment;
    }

    public void setUserEmail(String userEmail) {
        model.setEmail(userEmail);
    }

    public void setUserPassword(String userPassword) {
        model.setPassword(userPassword);
    }

    public void setUserFullName(String userFullName) {
        model.setFullName(userFullName);
    }

    public void setUserPhoneNumber(String userPhoneNumber) {
        model.setPhoneNumber(userPhoneNumber);
    }

    /**
     * Xử lý những thao tác sau khi nhận view.getId() từ View.onClickListener
     */
    public void handleOnClick(View view) {
        int id = view.getId();

        if (id == R.id.tvToRegister) {
            loginFragment.navigateToRegisterFragment();
        } else if (id == R.id.btnLogin) {
            loginFragment.setEnabledButton(false);
            login();
        } else if (id == R.id.layoutLoginFragment) {
            UiController.getInstance().closeKeyBoard(loginFragment.requireActivity());
        }

        if (id == R.id.tvToLogin) {
            registerFragment.navigateToLoginFragment();
        } else if (id == R.id.btnRegister) {
            register();
        } else if (id == R.id.layoutRegisterFragment) {
            UiController.getInstance().closeKeyBoard(registerFragment.requireActivity());
        }
    }

    /**
     * Lấy dữ liệu từ view và kiểm tra xem tài khoản người dùng cung cấp từ view có hợp lệ
     * hay không ? Nếu hợp lệ thì chuyển sang Trang làm việc chính, không thì hiển thị thông báo
     * đăng nhập thất bại
     */
    private void login() {
        UiController.getInstance().closeKeyBoard(loginFragment.requireActivity());
        UiController.getInstance().initProgressDialog(loginFragment.getContext());
        UiController.getInstance().turnOnLoading(loginFragment.getContext());
        clearErrorMessageLogin();
        loginFragment.getValue();
        if (model.isValidLoginUser() && UserRepository.isMatchedUser(model)) {
            DialogController.getInstance().showSuccessDialog(loginFragment.getContext(), id -> {
                loginFragment.navigateToMainActivity();
                loginFragment.setEnabledButton(true);
            }, ConstantMessage.MESSAGE_SUCCESS_LOGIN);
            UiController.getInstance().turnOffLoading();
            return;
        }
        if (model.isValidLoginUser()) {
            DialogController.getInstance().showErrorDialog(loginFragment.getContext(),
                    id -> loginFragment.setEnabledButton(true),
                    ConstantMessage.MESSAGE_ERROR_LOGIN);
            UiController.getInstance().turnOffLoading();
            return;
        }

        UiController.getInstance().turnOffLoading();

        if (model.getEmail().isEmpty()) {
            loginFragment.showErrorEmail(ConstantMessage.MESSAGE_EMPTY_EMAIL);
        } else if (!model.isValidEmail()) {
            loginFragment.showErrorEmail(ConstantMessage.MESSAGE_INVALID_EMAIL);
        } else {
            loginFragment.showErrorEmail(null);
        }

        if (model.getPassword().isEmpty()) {
            loginFragment.showErrorPassword(ConstantMessage.MESSAGE_EMPTY_PASSWORD);
        } else if (!model.isValidPassword()) {
            loginFragment.showErrorPassword(ConstantMessage.MESSAGE_INVALID_PASSWORD);
        } else {
            loginFragment.showErrorPassword(ConstantMessage.CLEAR_TEXT);
        }
        loginFragment.setEnabledButton(true);
    }

    private void clearErrorMessageLogin() {
        loginFragment.showErrorEmail(ConstantMessage.CLEAR_TEXT);
        loginFragment.showErrorPassword(ConstantMessage.CLEAR_TEXT);
    }

    private void clearErrorMessageRegister() {
        registerFragment.showErrorEmail(ConstantMessage.CLEAR_TEXT);
        registerFragment.showErrorPassword(ConstantMessage.CLEAR_TEXT);
        registerFragment.showErrorFullName(ConstantMessage.CLEAR_TEXT);
        registerFragment.showErrorPhoneNumber(ConstantMessage.CLEAR_TEXT);
    }

    /**
     * Lấy dữ liệu từ view và xử lý logic business để đăng ký tài khoản
     */
    private void register() {
        UiController.getInstance().closeKeyBoard(registerFragment.requireActivity());
        UiController.getInstance().initProgressDialog(registerFragment.getContext());
        UiController.getInstance().turnOnLoading(registerFragment.getContext());
        clearErrorMessageRegister();
        registerFragment.getValue();
        if (UserRepository.isExistedUser(model)) {
            UiController.getInstance().turnOffLoading();
            DialogController.getInstance().showErrorDialog(registerFragment.getContext(),
                    id -> registerFragment.setEnabledButton(true),
                    ConstantMessage.MESSAGE_EXISTS_EMAIL);
            return;
        }
        if (model.isValidUser()) {
            UserRepository.createUser(model);
            UiController.getInstance().turnOffLoading();
            DialogController.getInstance().showSuccessDialog(registerFragment.getContext(), id -> {
                registerFragment.navigateToLoginFragment();
                registerFragment.setEnabledButton(true);
            }, ConstantMessage.MESSAGE_SUCCESS_REGISTER);
            return;
        }
        UiController.getInstance().turnOffLoading();

        if (model.getEmail().isEmpty()) {
            registerFragment.showErrorEmail(ConstantMessage.MESSAGE_EMPTY_EMAIL);
        } else if (!model.isValidEmail()) {
            registerFragment.showErrorEmail(ConstantMessage.MESSAGE_INVALID_EMAIL);
        } else {
            registerFragment.showErrorEmail(ConstantMessage.CLEAR_TEXT);
        }

        if (model.getPassword().isEmpty()) {
            registerFragment.showErrorPassword(ConstantMessage.MESSAGE_EMPTY_PASSWORD);
        } else if (!model.isValidPassword()) {
            registerFragment.showErrorPassword(ConstantMessage.MESSAGE_INVALID_PASSWORD);
        } else {
            registerFragment.showErrorPassword(ConstantMessage.CLEAR_TEXT);
        }

        if (model.getFullName().isEmpty()) {
            registerFragment.showErrorFullName(ConstantMessage.MESSAGE_EMPTY_FULL_NAME);
        } else {
            registerFragment.showErrorFullName(ConstantMessage.CLEAR_TEXT);
        }

        if (model.getPassword().isEmpty()) {
            registerFragment.showErrorPhoneNumber(ConstantMessage.MESSAGE_EMPTY_PHONE_NUMBER);
        } else if (!model.isValidPhoneNumber()) {
            registerFragment.showErrorPhoneNumber(ConstantMessage.MESSAGE_INVALID_PHONE_NUMBER);
        } else {
            registerFragment.showErrorPhoneNumber(ConstantMessage.CLEAR_TEXT);
        }
        registerFragment.setEnabledButton(true);
    }
}

File: app/src/main/java/com/ptithcm/thuan6420/basecleanarchitecture/ui/login/UserController.kt
package com.ptithcm.thuan6420.basecleanarchitecture.ui.login

import android.util.Patterns
import com.ptithcm.thuan6420.basecleanarchitecture.data.repositories.UserRepository
import com.ptithcm.thuan6420.basecleanarchitecture.ui.ResponseUI
import com.ptithcm.thuan6420.basecleanarchitecture.ui.util.ConstantMessage
import com.ptithcm.thuan6420.basecleanarchitecture.ui.util.Credentials.REGEX_PHONE_NUMBER
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.util.regex.Pattern

class UserController {

    val NO_FIELD: Byte = 0
    val ID_EMAIL_FIELD: Byte = 1
    val ID_PASSWORD_FIELD: Byte = 2
    val ID_FULLNAME_FIELD: Byte = 3
    val ID_PHONENUMBER_FIELD: Byte = 4

    fun isLoginSuccess(email: String, password: String) : ResponseUI {
        if (!UserRepository().isMatchedUser(User(email, password))) {
            return ResponseUI(false, ConstantMessage.MESSAGE_ERROR_LOGIN)
        }
        return ResponseUI(true, ConstantMessage.MESSAGE_SUCCESS_LOGIN)
    }

    fun isRegisterSuccess(
        email: String, password: String,
        fullName: String, phoneNumber: String
    ): ResponseUI {
        val user = User(email, password, fullName, phoneNumber)
        if (UserRepository().isExistedUser(user)) {
            return ResponseUI(false, ConstantMessage.MESSAGE_EXISTS_EMAIL)
        }
        UserRepository().createUser(user)
        return ResponseUI(true, ConstantMessage.MESSAGE_SUCCESS_REGISTER)
    }

    fun isValidEmail(email: String): Boolean {
        return Patterns.EMAIL_ADDRESS.matcher(email).matches()
    }

    fun isValidPassword(password: String): Boolean {
        return password.length >= 9
    }

    fun isValidPhoneNumber(phoneNumber: String): Boolean {
        return Pattern.compile(REGEX_PHONE_NUMBER).matcher(phoneNumber).matches()
    }

    /**
     * Tìm input field điền giá trị không hợp lệ trên Login Screen
     * 0 - no field
     * 1 - field email
     * 2 - field password
     * send error message
     */
    fun findInvalidLoginFields(email: String, password: String): List<InvalidField> {
        val invalidFields = listOf(InvalidField(1, null), InvalidField(2, null))
        if (email.isEmpty()) {
            invalidFields.get(0).message = ConstantMessage.MESSAGE_EMPTY_EMAIL
        } else if (!isValidEmail(email)) {
            invalidFields.get(0).message = ConstantMessage.MESSAGE_INVALID_EMAIL
        } else {
            invalidFields.get(0).message = null
        }

        if (password.isEmpty()) {
            invalidFields.get(1).message = ConstantMessage.MESSAGE_EMPTY_PASSWORD
        } else if (!isValidPassword(password)) {
            invalidFields.get(1).message = ConstantMessage.MESSAGE_INVALID_PASSWORD
        } else {
            invalidFields.get(1).message = null
        }
        return invalidFields
    }

    /**
     * Tìm input field điền giá trị không hợp lệ trên Register Screen
     * 0 - no field
     * 1 - field email
     * 2 - field password
     * 3 - field full name
     * 4 - phone number
     * send error message
     */
    fun findInvalidRegisterFields(
        email: String, password: String,
        fullName: String, phoneNumber: String
    ): List<InvalidField> {
        val invalidFields = listOf(
            InvalidField(1, null),
            InvalidField(2, null),
            InvalidField(3, null),
            InvalidField(4, null)
        )

        if (email.isEmpty()) {
            invalidFields.get(0).message = ConstantMessage.MESSAGE_EMPTY_EMAIL
        } else if (!isValidEmail(email)) {
            invalidFields.get(0).message = ConstantMessage.MESSAGE_INVALID_EMAIL
        } else {
            invalidFields.get(0).message = null
        }

        if (password.isEmpty()) {
            invalidFields.get(1).message = ConstantMessage.MESSAGE_EMPTY_PASSWORD
        } else if (!isValidPassword(password)) {
            invalidFields.get(1).message = ConstantMessage.MESSAGE_INVALID_PASSWORD
        } else {
            invalidFields.get(1).message = null
        }

        if (fullName.isEmpty()) {
            invalidFields.get(2).message = ConstantMessage.MESSAGE_EMPTY_FULL_NAME
        } else {
            invalidFields.get(2).message = null
        }

        if (phoneNumber.isEmpty()) {
            invalidFields.get(3).message = ConstantMessage.MESSAGE_EMPTY_PHONE_NUMBER
        } else if (!isValidPhoneNumber(phoneNumber)) {
            invalidFields.get(3).message = ConstantMessage.MESSAGE_INVALID_PHONE_NUMBER
        } else {
            invalidFields.get(3).message = null
        }

        return invalidFields
    }

    companion object {
        private var Instance: UserController? = null
        operator fun invoke() = synchronized(this) {
            if (Instance != null) {
                Instance
            }
            Instance = UserController()
        }
    }
}